src/PrestaShopBundle/Form/Admin/Improve/International/Localization/LocalizationConfigurationType.php

```diff
@@ -0,0 +1,200 @@
+
+/**
+ * Class LocalizationConfigurationType is responsible for building 'Improve > International > Localization' page
+ * 'Configuration' form
+ */
+class LocalizationConfigurationType extends AbstractType
+{
+    /**
+     * @var LegacyContext
+     */
+    private $legacyContext;
+
+    /**
+     * @var LanguageDataProvider
+     */
+    private $languageDataProvider;
+
+    /**
+     * @var CountryDataProvider
+     */
+    private $countryDataProvider;
+
+    /**
+     * @var CurrencyDataProvider
+     */
+    private $currencyDataProvider;
+
+    /**
+     * @var TimezoneRepository
+     */
+    private $timezoneRepository;
+
+    /**
+     * @param LegacyContext $legacyContext
+     * @param LanguageDataProvider $languageDataProvider
+     * @param CountryDataProvider $countryDataProvider
+     * @param CurrencyDataProvider $currencyDataProvider
+     * @param TimezoneRepository $timezoneRepository
+     */
+    public function __construct(
+        LegacyContext $legacyContext,
+        LanguageDataProvider $languageDataProvider,
+        CountryDataProvider $countryDataProvider,
+        CurrencyDataProvider $currencyDataProvider,
+        TimezoneRepository $timezoneRepository
+    ) {
+        $this->languageDataProvider = $languageDataProvider;
+        $this->countryDataProvider = $countryDataProvider;
+        $this->legacyContext = $legacyContext;
+        $this->currencyDataProvider = $currencyDataProvider;
+        $this->timezoneRepository = $timezoneRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('default_language', ChoiceType::class, [
+                'choices' => $this->getLanguageChoices(),
+                'attr' => [
+                    'data-toggle' => 'select2',
+                    'data-minimumResultsForSearch' => '7',
+                ],
+            ])
+            ->add('detect_language_from_browser', SwitchType::class)
+            ->add('default_country', ChoiceType::class, [
+                'choices' => $this->getCountryChoices(),
+                'attr' => [
+                    'data-toggle' => 'select2',
+                    'data-minimumResultsForSearch' => '7',
+                ],
+            ])
+            ->add('detect_country_from_browser', SwitchType::class)
+            ->add('default_currency', ChoiceType::class, [
+                'choices' => $this->getCurrencyChoices(),
+                'attr' => [
+                    'data-toggle' => 'select2',
+                    'data-minimumResultsForSearch' => '7',
+                ],
+            ])
+            ->add('timezone', ChoiceType::class, [
+                'choices' => $this->getTimezoneChoices(),
+                'attr' => [
+                    'data-toggle' => 'select2',
+                    'data-minimumResultsForSearch' => '7',
+                ],
+            ])
+        ;
+    }
+
+    /**
```

Comment:
You don't need to inject the world in this formtype :)

What I suggest here
* move every private function to his own "Provider/Repository": each one of them MUST return an array
* so the constructor of this formType become `__construct(array $languageChoices, array $countryChoices, array $currencyChoices, array $currencyChoices, array $timezoneChoices) { /* ... */}`

In fact, it's ok to have private function in form types "if" you don't inject new dependencies only to be used in private functions: your private functions are about internal changes of your class: if you change something not related in it, this probably means this function should not be here :)

Reply:
hmm, should we have some generic interface under `PrestaShopBundle\Form\Admin` namespace like `ChoiceValueProviderInterface` with `getChoices(): array` method so it could be reused multiple times (in this PR i need 5 different choice providers).  what do you think @mickaelandrieu ?

Reply:
no, because I don't see why values of the form should be part of the contract.... and in this case, which "choices" should be in "getChoices"?  $languageChoices, $countryChoices, $currencyChoices or $timezoneChoices?

We usually rely on interface to protect an implementation from his dependencies, but here we inject the dependencies so we are already protected: we can already type hint on ``array``.  Am I missing something here?

Reply:
Oh! interesting :) This can formalize a constraint on ``getChoices`` ... can you add an issue on prestafony-project with this proposal? I have the feeling we should discuss it, but not in this pull request.

Still, I'm not sure *yet* of the gain to introduce the interface: does it make the code more testable, easier to understand or to extend? Are these arrays need to be strictly called "choices" or can they be used in other contexts?